wp-h5bp-htaccess.php

```diff
@@ -1,46 +1,113 @@
-    if ($mod_rewrite_enabled) {
-      $h5bp_rules = extract_from_markers($htaccess_file, 'HTML5 Boilerplate');
-      if ($h5bp_rules === array()) {
-        $filename = dirname(__FILE__) . '/h5bp-htaccess';
-        return insert_with_markers($htaccess_file, 'HTML5 Boilerplate', extract_from_markers($filename, 'HTML5 Boilerplate'));
-      }
+  protected static $mod_rewrite_enabled;
+  protected static $home_path;
+  protected static $wp_htaccess_file;
+  protected static $roots_htaccess_file;
+  
+  protected $errors = [];
+
+  public static function init() {
+    if (!self::$instance) {
+      self::$instance = new static();
+    }
+    return self::$instance;
+  }
+  
+  protected function __construct() {
+    require_once ABSPATH . '/wp-admin/includes/file.php';
+    require_once ABSPATH . '/wp-admin/includes/misc.php';
+    
+    self::$mod_rewrite_enabled = got_mod_rewrite();
+    self::$home_path = get_home_path();
+    self::$wp_htaccess_file = self::$home_path . '.htaccess';
+    self::$roots_htaccess_file = locate_template(['server_configs.conf','h5bp-htaccess.conf']);
+    if (!self::$roots_htaccess_file) {
+      self::$roots_htaccess_file = __DIR__ . '/h5bp-htaccess.conf';
+    }
+    
+    if (!$this->verifySetup()) {
+      $this->alerts();
+    } else {
+      self::setRulesFilters();
+      $this->write();
+    }
+  }
+  
+  public static function setRulesFilters() {
+    $home_url = parse_url(is_multisite() ? network_home_url('/') : home_url('/'));
+    extract($home_url);
+    self::$rules_filters = [
+      'www\\.example\\.com' => preg_quote($host),
+      'www.example.com/' => $host . $path,
+      'www.example.com'  => $host,
+      '/wordpress' => parse_url(site_url(), PHP_URL_PATH)
+    ];
+    self::$rules_filters = apply_filters('roots/h5bp-htaccess-filters', self::$rules_filters);
+  }
+  
+  protected function verifySetup() {
+    if (!get_option('permalink_structure')) {
+      $this->errors[] = sprintf(__('Please enable %s.', 'roots'), '<a href="' . admin_url('options-permalink.php') . '">Permalinks</a>');
+    }
+    
+    if (!$this->isWritable()) {
+      $this->errors[] = sprintf(__('Please make sure your %s file is writable.', 'roots'), '<a href="' . admin_url('options-permalink.php') . '">.htaccess</a>');
+    }
+    
+    if (!self::$mod_rewrite_enabled) {
+      $this->errors[] = sprintf(__('Please enable %s.', 'roots'), '<a target="_blank" href="http://httpd.apache.org/docs/current/rewrite/">Apache mod_rewrite</a>');
+    }
+    
+    if (!file_exists(self::$roots_htaccess_file)) {
+      $this->errors[] = sprintf(__('Cannot find the file %s.', 'roots'), self::$roots_htaccess_file);
+    }
+    
+    return empty($this->errors);
+  }
+  
+  protected function isWritable() {
+    return is_writable(self::$home_path) || file_exists(self::$wp_htaccess_file) && is_writable(self::$wp_htaccess_file);
```

Comment:
This could use some extra parantheses for grouping to make the intention clearer.